### PR TITLE
[ONNX] Update constant-folding of Gather op

### PR DESCRIPTION
**Pull request by KsenijaS**
Update constant-folding of Gather operator so it also includes cases where rank of indices input is 0.
Currently it only support cases where rank of indices is 1.

### CODE REVIEW

File: test/onnx/test_pytorch_onnx_onnxruntime.py [hunk 1]
```diff
@@ -2726,6 +2726,36 @@ def forward(self, input, indices):
         indices = torch.tensor([[1, 0], [0, 1], [0, 1]], dtype=torch.int64)
         self.run_test(GatherModel(), input=(input, indices))
 
+    @skipIfUnsupportedMinOpsetVersion(11)
+    def test_gather_constant_fold(self):
```
Comment (neginraoof) [inline test/onnx/test_pytorch_onnx_onnxruntime.py:2730]: Move these tests to utility tests?

Comment (neginraoof) [reply test/onnx/test_pytorch_onnx_onnxruntime.py:2730]: Maybe also describe each test is testing which case (rank 0, etc.) in comments.

Comment (KsenijaS) [reply test/onnx/test_pytorch_onnx_onnxruntime.py:2730]: utility tests only test if the node has been constant-folded, not if it has been done in the correct way. Here the issue is that constant_folding was outputting the incorrect result.

File: test/onnx/test_pytorch_onnx_onnxruntime.py [hunk 2]
```diff
@@ -2726,6 +2726,38 @@ def forward(self, input, indices):
...
+            def forward(self, x):
+                # shape is of rank 0
+                shape = self.weight.shape[0]
+                m = 5 - shape
+                return x.clamp(min=m)
+
+        x = torch.randn(1)
+        self.run_test(GatherModule(), (x,))
+
+        class GatherModule(torch.nn.Module):
```
Comment (jiafatom) [inline test/onnx/test_pytorch_onnx_onnxruntime.py:2745]: We have `class GatherModule` above. Use another name?

File: torch/csrc/jit/passes/onnx/constant_fold.cpp [hunk 3]
```diff
@@ -313,6 +313,16 @@ c10::optional<at::Tensor> runTorchBackendForOnnx(
...
     updated_val = at::index_select(inputTensorValues[0], axis, indices_masked);
+    auto q = indices.dim();
+    // Cases where rank of indices > 1 are not currently supported.
+    if (q > 1) {
+      return c10::nullopt;
+    }
+    // If rank of indices is 0, rank of output tensor should be
+    // rank_of_input - 1.
+    if (q < 1) {
+      updated_val = updated_val.squeeze();
```
Comment (jiafatom) [inline torch/csrc/jit/passes/onnx/constant_fold.cpp:324]: This is a good fix! My PR also depend on this change. Let's get this PR in asap. Thanks.

File: test/onnx/test_pytorch_onnx_onnxruntime.py [hunk 4]
```diff
@@ -2726,6 +2726,38 @@ def forward(self, input, indices):
...
+    @skipIfUnsupportedMinOpsetVersion(11)
+    def test_gather_constant_fold(self):
+        class GatherModule(torch.nn.Module):
+            def __init__(self):
+                super(GatherModule, self).__init__()
+                self.register_buffer("weight", torch.ones(5))
+
+            def forward(self, x):
+                # shape is of rank 0
+                shape = self.weight.shape[0]
```
Comment (jiafatom) [inline test/onnx/test_pytorch_onnx_onnxruntime.py:2738]: I don't see a torch gather there. Is this `shape[0]` converting to onnx Gather?

File: test/onnx/test_pytorch_onnx_onnxruntime.py (hunk 2 above)
Comment (KsenijaS) [reply test/onnx/test_pytorch_onnx_onnxruntime.py:2745]: yes, but it doesn't matter in this case

File: test/onnx/test_pytorch_onnx_onnxruntime.py (hunk 4 above)
Comment (KsenijaS) [reply test/onnx/test_pytorch_onnx_onnxruntime.py:2738]: yes shape[0] is converting to Gather op

**Review by jiafatom — APPROVED** [review]

File: test/onnx/test_pytorch_onnx_onnxruntime.py (hunk 1 above)
Comment (neginraoof) [reply test/onnx/test_pytorch_onnx_onnxruntime.py:2738]: Actually they could test both. Anyways, constant folding is a utility.

**Pull request merged**